### PR TITLE
Seesaw Neopixel Support

### PR DESCRIPTION
**Pull request by trichner**
As advertised in https://github.com/tinygo-org/drivers/issues/564

Most of it is refactored and goifyed from the offical Arduino driver: https://github.com/adafruit/Adafruit_Seesaw/blob/master/Adafruit_seesaw.cpp

This is support for the seesaw device as well as its usage with NeoPixels.

### CODE REVIEW

File: seesaw/seesaw.go [hunk 1]
```diff
@@ -0,0 +1,113 @@
+package seesaw
+
+import (
+	"errors"
+	"strconv"
```
Comment (trichner) [inline seesaw/seesaw.go]: Is there a more lightweight alternative to `strconv`?

File: seesaw/seesaw.go [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+	"time"
+)
+
+const DefaultSeesawAddress = 0x49
+
+// empirically determined delay, the one from the official library seems to be too short (250us)
+const defaultDelay = 10 * time.Millisecond
```
Comment (trichner) [inline seesaw/seesaw.go:11]: I'd be happy to have a better alternative. That's also how the official library does it :/

**Comment by aykevl** [comment]
If I understand correctly, this driver works with an external I2C device that actually sends out the WS2812 protocol?

We already have a ws2812 driver, perhaps this driver could match the interface? In particular, the `WriteColors` method.

**Comment by trichner** [comment]
> If I understand correctly, this driver works with an external I2C device that actually sends out the WS2812 protocol?
> 
> We already have a ws2812 driver, perhaps this driver could match the interface? In particular, the `WriteColors` method.

You are correct.

@aykevl it seems the current driver deals with RGBA colors discarding the A part. According to the datasheet NeoPixels are either RGB or RGBW. Some ideas:

1. add `WriteColorsRGB` and `WriteColorsRGBW` or similar
2. add something similar to color.RGBA for RGBW and have a single `WriteColors` (my preference) with maybe a convenience `WriteColorsRGB` accepting RGBA discarding the A

What are your thoughts?

For my own reference: https://github.com/tinygo-org/drivers/blob/release/ws2812/ws2812.go#L36
Also found: https://github.com/davidadeleon/gophercon2022Badge/blob/main/neopixel/neopixel.go

File: seesaw/neopixel/neopixel.go [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+	l := s.pixelType.PutRGBW(buf, color)
+	if l != encodedLen {
+		panic("unexpected encoded length: " + strconv.Itoa(l) + " != " + strconv.Itoa(encodedLen))
+	}
+	byteOffset := offset * uint16(encodedLen)
+	return s.writeBuffer(byteOffset, buf)
+}
+
+// WriteColors writes the given colors to the seesaws NeoPixel buffer
+func (s *Device) WriteColors(buf []RGBW) error {
```
Comment (trichner) [inline seesaw/neopixel/neopixel.go:104]: Similar to https://github.com/tinygo-org/drivers/blob/release/ws2812/ws2812.go#L36 but with RGBW (that's what the device supports).

**Comment by trichner** [comment]
@aykevl 
What else is needed to get a review/get this merged?

I'm eager to add more but don't want to make this PR any bigger ;)

**Comment by trichner** [comment]
@deadprogram may I request a review? :)

File: seesaw/seesaw.go [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+	time.Sleep(delay)
+
+	return d.bus.Tx(d.addr, nil, buf)
+}
+
+func (d *Device) Write(module ModuleBaseAddress, function FunctionAddress, buf []byte) error {
+	prefix := []byte{byte(module), byte(function)}
+	data := append(prefix, buf...)
+	return d.bus.Tx(d.addr, data, nil)
+}
```
Comment (aykevl) [inline seesaw/seesaw.go:126]: These look like internal methods, why did you export them?

File: seesaw/seesaw.go [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+		lastErr = err
+		time.Sleep(10 * time.Millisecond)
+	}
+
+	if lastErr != nil {
+		return fmt.Errorf("failed to read hardware ID after reset, tryed %d times: %w", tries, lastErr)
+	}
+
+	return nil
+}
```
Comment (aykevl) [inline seesaw/seesaw.go:59]: This also looks like an internal function.

File: seesaw/seesaw.go (hunk 2 above)
Comment (aykevl) [reply seesaw/seesaw.go:11]: Sounds like this is actually a bug in the Seesaw (I2C should not need a delay like this).
Seems fine by me, if this is what's needed.

File: seesaw/seesaw.go [hunk 6]
```diff
@@ -50,7 +52,7 @@ func (d *Device) Begin() error {
 	}
 
 	if lastErr != nil {
-		return lastErr
+		return fmt.Errorf("failed to read hardware ID after reset, tryed %d times: %w", tries, lastErr)
```
Comment (aykevl) [inline seesaw/seesaw.go:55]: Please don't use `fmt` in drivers, that package is far too big for some microcontrollers.

File: seesaw/seesaw.go [hunk 7]
```diff
@@ -67,7 +69,18 @@ func (d *Device) ReadHardwareID() (byte, error) {
...
+
+// ReadVersion reads the version bytes from the device (undocumented in the datasheet)
+func (d *Device) ReadVersion() (uint32, error) {
+
+	buf := make([]byte, 4)
+	err := d.Read(ModuleStatusBase, FunctionStatusVersion, buf, defaultDelay)
+	if err != nil {
+		return 0, err
+	}
+	return (uint32(buf[0]) << 24) | (uint32(buf[1]) << 16) | (uint32(buf[2]) << 8) | uint32(buf[3]), nil
```
Comment (aykevl) [inline seesaw/seesaw.go:83]: You have made some changes to the generic seesaw support (seesaw/seesaw.go), perhaps this should be included in #565?

**Review by aykevl — COMMENTED** [review]
I'm experimenting with different APIs in https://github.com/aykevl/board, and I've actually added addressable LEDs yesterday. It uses generics to abstract over different color formats.
I will make a proposal for a new API sometime in the future, but that's the future, and not very useful now for the Seesaw.

If you ask me, I'd actually recommend making the PR a lot smaller to only implement the basics. That is:

* `WriteColors` - but perhaps using `color.RGBA` to use the same interface as the ws2812 package.

And not this:

  * `PixelType` (just default to GRB for now, like the ws2812 driver).
  * `RGBW` type.

I'm not sure about `SpeedKHz800` - are people still using 400kHz? As far as I know, practically all WS2812-like LED strips are 800kHz by now. I recommend not including it, and adding it if there is demand for it (to reduce the API surface).

Maybe I'm going a bit too far though, I like to keep things simple. As they say: "no is temporary, yes is forever". It's easy to add new features at a later time if needed, it's very difficult to remove or change them once included.

**Comment by aykevl** [comment]
Oh and I forgot, can you also add an example to the examples directory, and a smoke test to smoketest.sh?

File: seesaw/seesaw.go (hunk 6 above)
Comment (trichner) [reply seesaw/seesaw.go:55]: Fair enough. I went through the other drivers and didn't find a decent solution for adding crucial information
to errors apart from straight byte-to-string conversions.

I can whip up a quick and dirty byte-to-hex formatter or just not add the context to the error.
What is the preferred solution?

File: seesaw/seesaw.go (hunk 5 above)
Comment (trichner) [reply seesaw/seesaw.go:59]: I'll clean it up and move it into the `SoftReset`, it's anyways just a 'send reset command and wait for boot' method.

File: seesaw/seesaw.go (hunk 4 above)
Comment (trichner) [reply seesaw/seesaw.go:126]: They are not internal, they are quite crucial to build drivers for the other modules ontop of it. An alternative would be to just dump everything into the same package, but most of the time driver users are interested in one or two of the seesaw modules but not the rest.

I think putting it all in the same package probably doesn't make sense.

Hence the Read/Writer as well as ReadRegister and WriteRegister are exposed.

**Comment by trichner** [comment]
First of, thanks a ton for your review and feedback! I'm still a bit new to tinygo and its drivers, so bear with me while I learn :)

In general regarding the structure that I envisoned:

1. package `seesaw` supporting the basic function Read/Write operations as well as reset. I'll remove the other bits, they are not necessary and can be added later.
2. package `seesaw/neopixel` abstracting the neopixel module on the seesaw, depends on seesaw package. Only implementing the pixel type & speed I need for 4. for now.
3. package `seesaw/keypad` abstracting the keypad module on the seesaw, depends on seesaw package
4. package `neotrellis`, which is a 4x4 keyboard driven by the seesaw chip using the `neopixel` and `keypad` module and depending on them

Overall I don't think having one package trying to support all modules and functions is helpful. It also stops others from implementing their own abstractions for the seesaw modules without either copy-ing the package or extending it here.

I think it makes sense to also structure my PRs that way, so I'll incorporate your feedback and split it into those four.


> If you ask me, I'd actually recommend making the PR a lot smaller to only implement the basics. That is:
>     WriteColors - but perhaps using color.RGBA to use the same interface as the ws2812 package.
> 
> And not this:
> 
>    PixelType (just default to GRB for now, like the ws2812 driver).
>    RGBW type.

Removing `PixelType` in favour of hardcoding a default sounds good to me. Regarding RGBW see below.

> I'm not sure about SpeedKHz800 - are people still using 400kHz? As far as I know, practically all WS2812-like LED strips are 800kHz by now. I recommend not including it, and adding it if there is demand for it (to reduce the API surface).

I'm not sure either, that's what the seesaw itself supports. But can be added later, will remove in favour of hardcoded 800KHz

> Maybe I'm going a bit too far though, I like to keep things simple. As they say: "no is temporary, yes is forever". It's easy to add new features at a later time if needed, it's very difficult to remove or change them once included.

There I agree very much :)

Though one thing regarding the `color.RGBA`: This is not a 'temporary no' but rather intentionally building an API that I can not extend to the full functionality of what the seesaw device can without breaking compatibility or adding another method.
Therefore I think it makes sense to have the complexity of the RGBW type while dropping all actual support for flexibility for now.

What do you think?

**Comment by trichner** [comment]
> Oh and I forgot, can you also add an example to the examples directory, and a smoke test to smoketest.sh?

Yes, shall do :)

**Comment by trichner** [comment]
I've finally come around to give it another go, apparently https://github.com/tinygo-org/tinygo/commit/9126b9573711b1990f80ca1bca564c4886b8f8f2 fixed my timing issues that I couldn't resolve. Once the next tinygo release is out I intend to give this another go.

**Comment by deadprogram** [comment]
Sounds great @trichner

**Comment by trichner** [comment]
As promised, taken up again in #612. Closing this.

**Pull request closed without merging**